server/handlers_test.go

```diff
@@ -48,3 +50,73 @@ func TestHandleHealthFailure(t *testing.T) {
 		t.Errorf("expected 500 got %d", rr.Code)
 	}
 }
+
+type emptyStorage struct {
+	storage.Storage
+}
+
+func (*emptyStorage) GetAuthRequest(string) (storage.AuthRequest, error) {
+	return storage.AuthRequest{}, storage.ErrNotFound
+}
+
+func TestHandleInvalidOAuth2Callbacks(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	httpServer, server := newTestServer(ctx, t, func(c *Config) {
+		c.Storage = &emptyStorage{c.Storage}
+	})
+	defer httpServer.Close()
+
+	tests := []struct {
+		TargetURI    string
+		ExpectedCode int
+	}{
+		{"/callback", http.StatusBadRequest},
+		{"/callback?code=&state=", http.StatusBadRequest},
+		{"/callback?code=AAAAAAA&state=BBBBBBB", http.StatusBadRequest},
+	}
+
+	rr := httptest.NewRecorder()
+
+	for i, r := range tests {
+		server.ServeHTTP(rr, httptest.NewRequest("GET", r.TargetURI, nil))
+		if rr.Code != r.ExpectedCode {
+			t.Fatalf("test %d expected %d, got %d", i, r.ExpectedCode, rr.Code)
```

Comment:
Ok, this is nitpicking, but we don't have to use Fatal here -- if one fails, the following ones could still be tested.
Also, if you use `t.Run("test name or number", func (t *testing) { /* test body */ })`, you _can_ use `t.Fatal` inside the subtest without failing the entire test.

At any rate this is better than what we have right now, so let's get this merged, and extend on it as needed. 😃